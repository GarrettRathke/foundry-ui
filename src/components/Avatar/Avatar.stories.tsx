import * as React from 'react';
import { Story, Meta } from '@storybook/react';
import { withFoundryContext } from '../../../.storybook/decorators';

import Avatar, { AvatarProps } from './Avatar';

export const DefaultAvatar: Story<AvatarProps> = (args: AvatarProps) => <Avatar {...args} />;
DefaultAvatar.args = {
  initials: 'AA',
  imgURL:
    'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSHCWDau1lAskHdiokbaocBqIXOCxWcdlAwg9UhusQHaALmDwbfaDxGpjmn7Cv0HWXYxTI&usqp=CAU',
  size: 3,
  shape: 50,
  isLoading: false,
  isError: false,
  hasImage: true,
};

export default {
  title: 'Avatar',
  component: Avatar,
  decorators: [withFoundryContext],
  argTypes: {
    shape: { control: { type: 'range', min: 0, max: 50, step: 1 } },
  },
} as Meta;
