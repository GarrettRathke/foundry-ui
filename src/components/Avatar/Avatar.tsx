import styled from 'styled-components';
import React, { ReactNode } from 'react';
import { StyledBaseDiv, StyledBaseSpan } from '../../htmlElements';
import { useTheme } from '../../context';
import { SubcomponentPropsType, StyledSubcomponentType } from '../commonTypes';
import Skeleton from '../Skeleton/Skeleton';
import colors from '../../enums/colors';
import { readableColor } from 'polished';

export type AvatarContainerProps = {
  size: number;
  borderRadiusPercent: number;
  imgURL?: string;
  containerColor: string;
};

export type AvatarTextProps = {
  size: number;
  textColor: string;
};

export type AvatarProps = {
  placeholder?: ReactNode | string;
  imgURL?: string;
  size?: number;
  containerColor?: string;
  borderRadiusPercent?: number;
  isLoading?: boolean;
  isError?: boolean;

  StyledAvatarContainer?: StyledSubcomponentType;
  StyledAvatarText?: StyledSubcomponentType;
  StyledLoadingContainer?: StyledSubcomponentType;

  avatarContainerProps?: SubcomponentPropsType;
  avatarTextProps?: SubcomponentPropsType;

  avatarContainerRef?: React.RefObject<HTMLDivElement>;
  avatarTextRef?: React.RefObject<HTMLSpanElement>;
  avatarLoadingRef?: React.RefObject<HTMLDivElement>;
};

export const AvatarContainer = styled(StyledBaseDiv)`
  ${({ size, borderRadiusPercent, imgURL, containerColor }: AvatarContainerProps) => {
    if (imgURL) {
      return `
        border-radius: ${borderRadiusPercent}%;
        overflow: hidden;
        width: ${size * 3}em;
        height: ${size * 3}em;
        background-image: url(${imgURL});
        background-size: cover;
      `;
    }
    return `
      display: flex;
      border-radius: ${borderRadiusPercent}%;
      background-color: ${containerColor};
      padding: 1em;
      width: ${size * 3}em;
      height: ${size * 3}em;
      justify-content: center;
      align-items: center;
    `;
  }}
`;

export const AvatarText = styled(StyledBaseSpan)`
  ${({ size, textColor }: AvatarTextProps) => {
    return `
      display: flex;
      padding: 0.66em; 
      color: ${textColor};
      font-size: ${size}em; 
      font-weight: 600;
      `;
  }};
`;

// export const LoadingContainer = styled(Skeleton.Container)`
//   ${({ borderRadiusPercent }: AvatarContainerProps) => {
//     return `
//         display: flex;
//         border-radius: ${borderRadiusPercent}%;
//         justify-content: center;
//         align-items: center;
//         overflow: hidden;
//       `;
//   }};
// `;

const Avatar = ({
  placeholder,
  imgURL,
  size = 3,
  borderRadiusPercent,
  containerColor,
  isLoading,
  isError,
  StyledAvatarContainer = AvatarContainer,
  StyledAvatarText = AvatarText,
  avatarContainerProps = {},
  avatarTextProps = {},
  avatarContainerRef,
  avatarTextRef,
}: AvatarProps): JSX.Element => {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const fontColor = readableColor(containerColor!, colors.grayMedium, colors.background);
  // if (isLoading) {
  //   return (
  //     <StyledLoadingContainer
  //       size={size}
  //       borderRadiusPercent={borderRadiusPercent}
  //       ref={avatarLoadingRef}
  //       {...avatarLoadingProps}
  //     >
  //       <Skeleton isLoading StyledContainer={StyledLoadingContainer}>
  //         <StyledAvatarContainer
  //           ref={avatarContainerRef}
  //           size={size}
  //           borderRadiusPercent={borderRadiusPercent}
  //           {...avatarContainerProps}
  //         />
  //       </Skeleton>
  //     </StyledLoadingContainer>
  //   );
  // }

  if (isError) {
    return (
      <StyledAvatarContainer
        ref={avatarContainerRef}
        size={size}
        borderRadiusPercent={borderRadiusPercent}
        imgURL=""
        {...avatarContainerProps}
      >
        <StyledAvatarText
          ref={avatarTextRef}
          size={size * 2}
          color={colors.destructive}
          {...avatarTextProps}
        >
          !
        </StyledAvatarText>
      </StyledAvatarContainer>
    );
  }

  return (
    <StyledAvatarContainer
      ref={avatarContainerRef}
      size={size}
      borderRadiusPercent={borderRadiusPercent}
      imgURL={imgURL}
      containerColor={containerColor}
      {...avatarContainerProps}
    >
      {!imgURL ? (
        <StyledAvatarText
          {...avatarTextProps}
          ref={avatarTextRef}
          size={size}
          textColor={fontColor}
        >
          {placeholder}
        </StyledAvatarText>
      ) : (
        ''
      )}
    </StyledAvatarContainer>
  );
};

export default Avatar;
