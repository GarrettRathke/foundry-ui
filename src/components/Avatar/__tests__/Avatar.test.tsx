import React from 'react';
import Avatar from '../Avatar';
import { render, waitFor, configure } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';

expect.extend(toHaveNoViolations);
configure({ testIdAttribute: 'data-test-id' });
const testId = 'foundry-avatar';

describe('Avatar', () => {
  it('matches snapshot', async() => {
    const { container } = render(<Avatar />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    })
  })
  it('matches snapshot Loading State', async () => {
    const { container } = render(<Avatar isLoading size={64} data-test-id={testId} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('matches snapshot Error State', async () => {
    const { container } = render(<Avatar isError size={64} data-testid={testId} />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar displays correct initials when their is no image', async () => {
    const { container } = render(
      <Avatar initials="MI" hasImage={false} size={64} data-testid={testId} />,
    );

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar displays correct image when their is an image', async () => {
    const { container } = render(
      <Avatar imgURL="https://tinyurl.com/49dba3d4"
       size={64} data-testid={testId} />,
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('avatar size is correct', async () => {
    const { container } = render(
      <Avatar size={21} data-test-id={testId} />,
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    })
  });

  it('avatar shape is correct', async () => {
    const { container } = render(
      <Avatar shape={30} data-test-id={testId} />,
    );
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    })
  });

  describe('Accessibility Tests', () => {
    it('Should pass accessibility test with default props', async () => {
      const component = <Avatar></Avatar>;
      const { container } = render(component);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  }); 

  describe('Ref Tests', () => {
    it('avatarContainer.current should exist', async () => {
      // const ref = React.createRef<HTMLDivElement>();
      // const { getByTestId } = render(
      //   <Avatar avatarContainerRef={ref} avatarProps={{ 'data-test-id': testId }} />,
      // );
      // await waitFor(() => getByTestId(testId));
      // expect(ref.current instanceof HTMLDivElement).toBeTruthy();
    });

    it('avatarText.current should exist', async () => {

    });


    it('avatarImage.current should exist', async () => {

    });


    it('avatarLoading.current should exist', async () => {

    });
  });
});
