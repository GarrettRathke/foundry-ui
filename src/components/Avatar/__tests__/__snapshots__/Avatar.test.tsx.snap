// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar avatar displays correct image when their is an image 1`] = `
.c0 {
  border-radius: undefined%;
  overflow: hidden;
  width: 192em;
  height: 192em;
  background-image: url(https://tinyurl.com/49dba3d4);
  background-size: cover;
}

<div>
  <div
    class=" c0"
    size="64"
  />
</div>
`;

exports[`Avatar avatar displays correct initials when their is no image 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: undefined%;
  background-color: undefined;
  padding: 1em;
  width: 192em;
  height: 192em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.66em;
  color: #51575D;
  font-size: 64em;
  font-weight: 600;
}

<div>
  <div
    class=" c0"
    size="64"
  >
    <span
      class=" c1"
      size="64"
    />
  </div>
</div>
`;

exports[`Avatar avatar shape is correct 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 30%;
  background-color: undefined;
  padding: 1em;
  width: 9em;
  height: 9em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.66em;
  color: #51575D;
  font-size: 3em;
  font-weight: 600;
}

<div>
  <div
    class=" c0"
    size="3"
  >
    <span
      class=" c1"
      size="3"
    />
  </div>
</div>
`;

exports[`Avatar avatar size is correct 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: undefined%;
  background-color: undefined;
  padding: 1em;
  width: 63em;
  height: 63em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.66em;
  color: #51575D;
  font-size: 21em;
  font-weight: 600;
}

<div>
  <div
    class=" c0"
    size="21"
  >
    <span
      class=" c1"
      size="21"
    />
  </div>
</div>
`;

exports[`Avatar matches snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: undefined%;
  background-color: undefined;
  padding: 1em;
  width: 9em;
  height: 9em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.66em;
  color: #51575D;
  font-size: 3em;
  font-weight: 600;
}

<div>
  <div
    class=" c0"
    size="3"
  >
    <span
      class=" c1"
      size="3"
    />
  </div>
</div>
`;

exports[`Avatar matches snapshot Error State 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: undefined%;
  background-color: undefined;
  padding: 1em;
  width: 192em;
  height: 192em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.66em;
  color: undefined;
  font-size: 128em;
  font-weight: 600;
}

<div>
  <div
    class=" c0"
    size="64"
  >
    <span
      class=" c1"
      color="#d00"
      size="128"
    >
      !
    </span>
  </div>
</div>
`;

exports[`Avatar matches snapshot Loading State 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: undefined%;
  background-color: undefined;
  padding: 1em;
  width: 192em;
  height: 192em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.66em;
  color: #51575D;
  font-size: 64em;
  font-weight: 600;
}

<div>
  <div
    class=" c0"
    size="64"
  >
    <span
      class=" c1"
      size="64"
    />
  </div>
</div>
`;
